File: testerum-backend/testerum-runner/testerum-runner-api/src/main/kotlin/com/testerum/runner/report_model/ReportTest.kt
package com.testerum.runner.report_model

import com.fasterxml.jackson.annotation.JsonProperty
import com.testerum_api.testerum_steps_api.test_context.ExecutionStatus
import java.time.LocalDateTime

data class ReportTest(@JsonProperty("testName")         val testName: String,
                      @JsonProperty("testFilePath")     val testFilePath: String,
                      @JsonProperty("tags")             val tags: List<String>,
                      @JsonProperty("startTime")        val startTime: LocalDateTime,
                      @JsonProperty("endTime")          val endTime: LocalDateTime,
                      @JsonProperty("durationMillis")   val durationMillis: Long,
                      @JsonProperty("status")           val status: ExecutionStatus,
                      @JsonProperty("textLogFilePath")  override val textLogFilePath: String,
                      @JsonProperty("modelLogFilePath") override val modelLogFilePath: String,
                      @JsonProperty("children")         val children: List<RunnerReportNode>) : FeatureOrTestRunnerReportNode

File: testerum-backend/testerum-runner/testerum-runner-cmdline/src/main/assembly/report_templates/common/testerum-model/report-model/model/report/report-test.ts
import {FeatureOrTestRunnerReportNode} from "./feature-or-test-runner-report-node";
import {ExecutionStatus} from "./execution-status";
import {ReportStep} from "./report-step";
import {MarshallingUtils} from "../../../json-marshalling/marshalling-utils";
import {RunnerReportNode, RunnerReportNodeType} from "./runner-report-node";
import {ReportHooks} from "./report-hooks";

export class ReportTest implements FeatureOrTestRunnerReportNode {

    constructor(public readonly testName: string,
                public readonly testFilePath: string,
                public readonly tags: Array<string>,
                public readonly startTime: Date,
                public readonly endTime: Date,
                public readonly durationMillis: number,
                public readonly status: ExecutionStatus,
                public readonly textLogFilePath: string,
                public readonly modelLogFilePath: string,
                public readonly children: Array<RunnerReportNode>) {}

    static parse(input: Object): ReportTest {
        if (!input) {
            return null;
        }

        const testName = input["testName"];
        const testFilePath = input["testFilePath"];
        const tags = MarshallingUtils.parseListOfStrings(input["tags"]);
        const startTime = MarshallingUtils.parseUtcToLocalDateTime(input["startTime"]);
        const endTime = MarshallingUtils.parseUtcToLocalDateTime(input["endTime"]);
        const durationMillis = input["durationMillis"];
        const status = MarshallingUtils.parseEnum(input["status"], ExecutionStatus);
        const textLogFilePath = input["textLogFilePath"];
        const modelLogFilePath = input["modelLogFilePath"];
        const children = MarshallingUtils.parseListPolymorphically<RunnerReportNode>(input["children"], {
            [RunnerReportNodeType[RunnerReportNodeType.STEP]]: ReportStep,
            [RunnerReportNodeType[RunnerReportNodeType.REPORT_HOOKS]]: ReportHooks
        });

        return new ReportTest(testName, testFilePath, tags, startTime, endTime, durationMillis, status, textLogFilePath, modelLogFilePath, children);
    }
}
